lib/libutils/ext/include/printk.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2015 Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * This file provides extensions to the standard snprintf() and vsnprintf()
+ * functions. These 'k' variants support additional formats when they are
+ * compiled for the kernel (i.e., when __KERNEL__ is defined).
+ * They do *not* provide any such extensions when build for user-mode -- in
+ * this case they are expected to behave like the standard ISO C functions.
+ */
+
+#ifndef PRINTK_H
+#define PRINTK_H
+
+#include <stddef.h>
+#include <stdarg.h>
+#include <stdbool.h>
+
+int snprintk(char *str, size_t size, const char *fmt, ...)
+		    __attribute__((__format__(__printf__, 3, 4)));
+int vsnprintk(char *str, size_t size, const char *fmt, va_list ap)
+		    __attribute__((__format__(__printf__, 3, 0)));
+
+int _vsnprintk(char *str, size_t size, const char *fmt, va_list ap,
```

Comment:
It's just that IMO `[v]snprintf` should really not have the extension because it is a deviation from the standard. If we add other custom formats, it might become a problem. So, `_vnsprintk` is used internally only, it just has to be visible from the wrappers in `snprintf.c`. I could declare it extern in there, instead.
Actually my first approach was to say: kernel has `[v]snprintk` only, and user mode has `[v]snprintf` only. But I found it a bit restrictive on the kernel side (we do have a few `snprintf` calls, why forbid them?). Similarly I thought it would not hurt to export the `[v]snprintk` versions to user mode, since it is in the `ext/` part of `libutils`. Hence this runtime choice.